Mathlib/Dynamics/TopologicalEntropy/CoverEntropy.lean

```diff
@@ -0,0 +1,590 @@
+/-
+Copyright (c) 2024 Damien Thomine. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damien Thomine, Pietro Monticone
+-/
+import Mathlib.Analysis.SpecialFunctions.Log.ENNRealLog
+import Mathlib.Data.Real.ENatENNReal
+import Mathlib.Dynamics.TopologicalEntropy.DynamicalEntourage
+
+/-!
+# Topological entropy via covers
+We implement Bowen-Dinaburg's definitions of the topological entropy, via covers.
+
+All is stated in the vocabulary of uniform spaces. For compact spaces, the uniform structure
+is canonical, so the topological entropy depends only on the topological structure. This will give
+a clean proof that the topological entropy is a topological invariant of the dynamics.
+
+A notable choice is that we define the topological entropy of a subset `F` of the whole space.
+Usually, one defines the entropy of an invariant subset `F` as the entropy of the restriction of the
+transformation to `F`. We avoid the latter definition as it would involve frequent manipulation of
+subtypes. Our version directly gives a meaning to the topological entropy of a subsystem, and a
+single theorem (`subset_restriction_entropy` in `TopologicalEntropy.Morphism`) will give the
+equivalence between both versions.
+
+Another choice is to give a meaning to the entropy of `∅` (it must be `-∞` to stay coherent) and to
+keep the possibility for the entropy to be infinite. Hence, the entropy takes values in the extended
+reals `[-∞, +∞]`. The consequence is that we use `ℕ∞`, `ℝ≥0∞` and `EReal` numbers.
+
+## Main definitions
+- `IsDynCoverOf`: property that dynamical balls centered on a subset `s` cover a subset `F`.
+- `coverMincard`: minimal cardinal of a dynamical cover. Takes values in `ℕ∞`.
+- `coverEntropyInfUni`/`coverEntropySupUni`: exponential growth of `coverMincard`. The former is
+defined with a `liminf`, the later with a `limsup`. Take values in `EReal`.
+- `coverEntropyInf`/`coverEntropySup`: supremum of `coverEntropyInfUni`/`coverEntropySupUni` over
+all entourages (or limit as the entourages go to the diagonal). These are Bowen-Dinaburg's
+versions of the topological entropy with covers. Take values in `EReal`.
+
```

Comment:
You need to mention also the main theorems here, notably the submultiplicativity one at finite times, and the fact that `coverEntropInf = coverEntropySup` when `F` is invariant.

Reply:
Done. I've also added `IsDynCoverOf.coverEntropyEnt_le_log_card_div`, which is of great practical use.